[BUG] `MissingArgumentException` is never thrown when a required agent argument is missing

**LangChain4j version:** `1.12.2`
  **Module:** `langchain4j-agentic`

  ### Description

  When invoking a sequence of agents and a required template variable is missing from the input map, the error handler receives an `AgentInvocationException`, but the exception is **not** an instance of `MissingArgumentException` — making it impossible to detect and recover from this specific failure case inside the error handler.

  According to the API design, the pattern shown in the documentation suggests checking `instanceof MissingArgumentException`:

  ```java
  .errorHandler(errorContext -> {
      AgentInvocationException exception = errorContext.exception();
      if (exception instanceof MissingArgumentException missingArgumentException) {
          // provide missing value and retry — this branch is NEVER reached
          missingArgumentException.agenticScope().writeState("topic", "default value");
          return ErrorRecoveryResult.retry();
      }
      return ErrorRecoveryResult.throwException();
  })
  ```

  ### Steps to Reproduce

  ```java

import dev.langchain4j.agentic.Agent;
import dev.langchain4j.agentic.AgenticServices;
import dev.langchain4j.agentic.UntypedAgent;
import dev.langchain4j.agentic.agent.AgentInvocationException;
import dev.langchain4j.agentic.agent.ErrorRecoveryResult;
import dev.langchain4j.agentic.agent.MissingArgumentException;
import dev.langchain4j.agentic.scope.AgenticScope;
import dev.langchain4j.model.chat.ChatModel;
import dev.langchain4j.model.googleai.GoogleAiGeminiChatModel;
import dev.langchain4j.service.UserMessage;
import dev.langchain4j.service.V;
import io.github.jtpadilla.example.format.Format;
import io.helidon.config.Config;

import java.util.Map;

public class AgentDemo {

    final static private String MODEL = "gemini-3.1-flash-lite-preview";

    final static private String API_KEY = Config.global().get("gemini-api-key").asString().orElseThrow(
            () -> new IllegalStateException("Configuration key 'gemini-api-key' is required"));

    public interface CreativeWriter {
        @UserMessage("""
            Eres un escritor creativo.
            Genera un borrador de una historia de no más de
            3 frases sobre el tema indicado.
            Devuelve solo la historia y nada más.
            El tema es {{topic}}.
            """)
        @Agent("Genera una historia basada en el tema indicado")
        String generateStory(@V("topic") String topic);
    }

    public interface AudienceEditor {
        @UserMessage("""
            Eres un editor profesional.
            Analiza y reescribe la siguiente historia para adaptarla mejor
            al público objetivo de {{audience}}.
            Devuelve solo la historia y nada más.
            La historia es "{{story}}".
            """)
        @Agent("Edita una historia para adaptarla mejor al público indicado")
        String editStory(@V("story") String story, @V("audience") String audience);
    }

    public interface StyleEditor {
        @UserMessage("""
            Eres un editor profesional.
            Analiza y reescribe la siguiente historia para que encaje mejor y sea más coherente con el estilo {{style}}.
            Devuelve solo la historia y nada más.
            La historia es "{{story}}".
            """)
        @Agent("Edita una historia para adaptarla mejor al estilo indicado")
        String editStory(@V("story") String story, @V("style") String style);
    }

    public interface NovelCreator {
        @Agent
        String createNovel(@V("topic") String topic, @V("audience") String audience, @V("style") String style);
    }

    public static void main(String[] args) {

        ChatModel chatModel = GoogleAiGeminiChatModel.builder()
                .apiKey(API_KEY)
                .modelName(MODEL)
                .logRequestsAndResponses(true)
                .build();

        UntypedAgent creativeWriter = AgenticServices.agentBuilder()
                .chatModel(chatModel)
                .name("creativeWriter")
                .description("Genera una historia basada en el tema indicado")
                .userMessage("""
                Eres un escritor creativo.
                Genera un borrador de una historia de no más de
                3 frases sobre el tema indicado.
                Devuelve solo la historia y nada más.
                El tema es {{topic}}.
                """)
                .inputKey(String.class, "topic")
                .returnType(String.class) // String is the default return type for untyped agents
                .outputKey("story")
                .build();

        UntypedAgent audienceEditor = AgenticServices.agentBuilder()
                .chatModel(chatModel)
                .name("audienceEditor")
                .description("Edita una historia para adaptarla mejor al público indicado")
                .userMessage("""
                Eres un editor profesional.
                Analiza y reescribe la siguiente historia para adaptarla mejor
                al público objetivo de {{audience}}.
                Devuelve solo la historia y nada más.
                La historia es "{{story}}".
                """)
                .inputKey(String.class, "audience")
                .inputKey(String.class, "story")
                .returnType(String.class) // String is the default return type for untyped agents
                .outputKey("story")
                .build();


        UntypedAgent styleEditor = AgenticServices.agentBuilder()
                .chatModel(chatModel)
                .name("styleEditor")
                .description("Edita una historia para adaptarla mejor al estilo indicado")
                .userMessage("""
                Eres un editor profesional.
                Analiza y reescribe la siguiente historia para que encaje mejor y sea más coherente con el estilo {{style}}.
                Devuelve solo la historia y nada más.
                La historia es "{{story}}".
                """)
                .inputKey(String.class, "style")
                .inputKey(String.class, "story")
                .returnType(String.class) // String is the default return type for untyped agents
                .outputKey("story")
                .build();

        UntypedAgent novelCreator = AgenticServices
                .sequenceBuilder()
                .subAgents(creativeWriter, audienceEditor, styleEditor)
                .outputKey("story")
                .errorHandler(errorContext -> {

                    // Se extraen los campos de ErrorContext
                    final AgentInvocationException exception = errorContext.exception();

                    // Se verifica si es el caso de prueba
                    if (exception instanceof MissingArgumentException missingArgumentException) {
                        // Es un Bug? No entra por aqui..
                        if (missingArgumentException.argumentName().equals("topic")) {
                            errorContext.agenticScope().writeState("topic", "ciencia ficcion galactica");
                            return ErrorRecoveryResult.retry();
                        }
                    }

                    // "dev.langchain4j.agentic.agent.AgentInvocationException: Failed to invoke agent method: public abstract java.lang.Object dev.langchain4j.agentic.UntypedAgent.invoke(java.util.Map)"
                    return ErrorRecoveryResult.throwException();

                })
                .build();

        Map<String, Object> input = Map.of(
                //"topic", "ciencia ficcion galactica",
                "style", "fantasía",
                "audience", "infantil"
        );

        String story = (String) novelCreator.invoke(input);
        System.out.println(Format.markdown(story));

    }

}

  ```

  ### Actual Behavior

  The application crashes with the following exception chain:

  ```
  Exception in thread "main" dev.langchain4j.agentic.agent.AgentInvocationException:
    Failed to invoke agent method: public abstract java.lang.Object dev.langchain4j.agentic.UntypedAgent.invoke(java.util.Map)
      at dev.langchain4j.agentic.internal.AgentInvoker.internalInvoke(AgentInvoker.java:43)
      ...
  Caused by: java.lang.reflect.InvocationTargetException
      ...
  Caused by: java.lang.reflect.UndeclaredThrowableException
      ...
  Caused by: java.lang.reflect.InvocationTargetException
      ...
  Caused by: java.lang.IllegalArgumentException: Value for the variable 'topic' is missing
      at dev.langchain4j.internal.Exceptions.illegalArgument(Exceptions.java:25)
      at dev.langchain4j.model.input.DefaultPromptTemplateFactory$DefaultTemplate
           .ensureAllVariablesProvided(DefaultPromptTemplateFactory.java:80)
      at dev.langchain4j.model.input.DefaultPromptTemplateFactory$DefaultTemplate
           .render(DefaultPromptTemplateFactory.java:60)
      at dev.langchain4j.model.input.PromptTemplate.apply(PromptTemplate.java:128)
      at dev.langchain4j.service.DefaultAiServices.prepareUserMessage(DefaultAiServices.java:574)
      ...
  ```

  ### Root Cause Analysis

  There are **two problems** in the agentic framework:

  1. **`MissingArgumentException` is never thrown.** The root cause is an `IllegalArgumentException` thrown by `DefaultPromptTemplateFactory#ensureAllVariablesProvided`, not a `MissingArgumentException`. The `langchain4j-core` layer does not know about (or does not use) the
  `MissingArgumentException` type from `langchain4j-agentic`.

  2. **The cause chain is not unwrapped.** The `AgentInvocationException` returned by `errorContext.exception()` wraps the root cause through three layers of reflection exceptions (`InvocationTargetException → UndeclaredThrowableException → InvocationTargetException`). Even if the root cause were
  a `MissingArgumentException`, the `instanceof` check on the top-level exception would still fail.

  ### Expected Behavior

  Either:
  - **Option A:** `DefaultPromptTemplateFactory#ensureAllVariablesProvided` should throw `MissingArgumentException` instead of `IllegalArgumentException`; AND the agentic error handling infrastructure should unwrap reflection wrappers before calling the error handler.
  - **Option B:** `AgentInvoker` / the error handling pipeline should detect the root cause in the chain and surface it as a `MissingArgumentException` when a missing variable is the underlying reason.

  This would allow the error handler to correctly intercept missing-argument scenarios and apply recovery logic (e.g., supply a default value and retry).

  ### Environment

  - Java: 21
  - LangChain4j: 1.12.2
  - Model: Google Gemini via `langchain4j-google-ai-gemini`

  ---
Bug Summary: MissingArgumentException never reaches the errorHandler because:

1. DefaultPromptTemplateFactory throws IllegalArgumentException (not MissingArgumentException) when a template variable is missing.

2. That exception is buried under three layers of reflection (InvocationTargetException → UndeclaredThrowableException → InvocationTargetException) before reaching the AgentInvocationException that the handler receives.

3. Therefore, the instance of MissingArgumentException never matches.

--- a/langchain4j-agentic/src/main/java/dev/langchain4j/agentic/internal/UntypedAgentInvoker.java
+++ b/langchain4j-agentic/src/main/java/dev/langchain4j/agentic/internal/UntypedAgentInvoker.java
@@ -1,5 +1,7 @@
 package dev.langchain4j.agentic.internal;
 
+import dev.langchain4j.agentic.agent.MissingArgumentException;
+import dev.langchain4j.agentic.planner.AgentArgument;
 import dev.langchain4j.agentic.scope.AgenticScope;
 import java.lang.reflect.Method;
 
@@ -11,6 +13,11 @@ public final class UntypedAgentInvoker extends AbstractAgentInvoker {
 
     @Override
     public AgentInvocationArguments toInvocationArguments(AgenticScope agenticScope) {
+        for (AgentArgument arg : arguments()) {
+            if (agenticScope.readState(arg.name()) == null) {
+                throw new MissingArgumentException(arg.name());
+            }
+        }
         return new AgentInvocationArguments(agenticScope.state(), new Object[]{agenticScope.state()});
     }
 

--- a/langchain4j-agentic/src/test/java/dev/langchain4j/agentic/WorkflowAgentsIT.java
+++ b/langchain4j-agentic/src/test/java/dev/langchain4j/agentic/WorkflowAgentsIT.java
@@ -314,13 +314,37 @@ public class WorkflowAgentsIT {
     }
 
     @Test
-    void sequential_agents_with_error_recovery_tests() {
+    void sequential_declarative_agents_with_error_recovery_tests() {
+        sequential_agents_with_error_recovery_tests(false);
+    }
+
+    @Test
+    void sequential_programmatic_agents_with_error_recovery_tests() {
+        sequential_agents_with_error_recovery_tests(true);
+    }
+
+    void sequential_agents_with_error_recovery_tests(boolean useProgrammaticAgent) {
         AtomicBoolean errorRecoveryCalled = new AtomicBoolean(false);
 
-        CreativeWriter creativeWriter = spy(AgenticServices.agentBuilder(CreativeWriter.class)
+        Object creativeWriter = useProgrammaticAgent ?
+            AgenticServices.agentBuilder()
+                .chatModel(baseModel())
+                .name("generateStory")
+                .description("Generate a story based on the given topic")
+                .userMessage("""
+                    You are a creative writer.
+                    Generate a draft of a story long no more than 3 sentence around the given topic.
+                    Return only the story and nothing else.
+                    The topic is {{topic}}.
+                    """)
+                .inputKey(String.class, "topic")
+                .returnType(String.class) // String is the default return type for untyped agents
+                .outputKey("story")
+                .build() :
+            AgenticServices.agentBuilder(CreativeWriter.class)
                 .chatModel(baseModel())
                 .outputKey("story")
-                .build());
+                .build();
 
         AudienceEditor audienceEditor = spy(AgenticServices.agentBuilder(AudienceEditor.class)
                 .chatModel(baseModel())
@@ -358,7 +382,6 @@ public class WorkflowAgentsIT {
 
         assertThat(errorRecoveryCalled.get()).isTrue();
 
-        verify(creativeWriter).generateStory("dragons and wizards");
         verify(audienceEditor).editStory(any(), eq("young adults"));
         verify(styleEditor).editStory(any(), eq("fantasy"));
     }

